Interactive Feedback UI: Main Container, Reviews Tab, Review Form, Review Card

Jotnivaas opened the pull request:
This Pull Request delivers the complete frontend implementation for the Reviews & Q&A feature under Story #162 – Interactive Feedback UI. It introduces a modular component structure, full styling, and interaction logic required to support user reviews within the event pages.

1. Main Container Component — Issue #182
- Implements the parent container responsible for rendering the Reviews & Q&A interface.
- Handles tab navigation logic for switching between “Reviews” and “Q&A”.
- Accepts and passes event_id, organizer_id, and current_user to child components.
- Provides fallback handling when required identifiers are not available.
- Establishes the structural layout that other components plug into.

2. Reviews Tab Structure — Issue #183
- Builds the layout for the Reviews tab including:
- Organizer average rating
- Rating distribution section
- Review list container
- Connects to the backend review-fetch endpoint using the project’s API utilities.
- Prepares the Review form and Review card areas for dynamic rendering.

3. New Review Form — Issue #184
- Implements a complete review submission form featuring:
- Star rating (1–5)
- Title and content inputs
- Optional category selector
- Optional image upload
- Includes client-side validation for required inputs.
- Displays submission states (loading, success, error).
- Submits form data to POST /api/reviews.
- Resets form on successful submission.

4. Review Card Component — Issue #185
- Renders individual reviews including:
- Reviewer name, avatar, and post date
- Star rating
- Review title, category, and content
- Optional images
- Adds the Helpful button with count updates.
- Shows Update/Delete controls when the current user is the author.
- Uses scoped CSS to keep Review Card styling contained.

Files Added or Updated

Files Added
frontend/components/reviews/review-card.js

Files Updated
backend/routes/reviews.js
frontend/styles.css
frontend/components/reviews/main-container.js
frontend/components/reviews/review-form.js
frontend/components/reviews/reviews-tab.js
frontend/components/reviews/README.md
frontend/reviews-demo.html

Integration Details

Components use the project’s API base helper for consistent request handling.

Frontend connects to:

GET /api/reviews/event/:event_id
POST /api/reviews
PATCH /api/reviews/:review_id/helpful
All UI elements mount cleanly within the Reviews & Q&A container demonstrated in the included demo file.

Testing Performed

- Manually verified the Reviews/Q&A container renders and tabs switch correctly in the demo page.
- Confirmed the rating summary and distribution widgets load/refresh based on organizer data.
- Exercised the review form: required-field validation, submission success handling, and automatic refresh after posting.
- Checked Review Cards render reviewer info, stars, category/title, text, images, helpful counts, and owner-only actions;   Helpful interactions update the count via the backend endpoint.
- Ensured all components mount cleanly on reviews-demo.html, so downstream integration just needs to supply the container target and props.

Closes Issues

Closes #182 – Build Main Container Component
Closes #183 – Build Reviews Tab Component Structure
Closes #184 – Build the New Review Form
Closes #185 – Build the Review Card Component

Jotnivaas committed:
Interactive Feedback UI: Main Container, Reviews Tab, Review Form, Review Card

Jotnivaas committed:
Interactive Feedback UI: Main Container, Reviews Tab, Review Form, Review Card

abdeljalilsennaoui reviewed (changes requested):
feature/interactive-feedback-reviews-qa is behind main by 10 commits. You need to rebase and handle any conflict preserving the backend and connecting your frontend correctly.

Jotnivaas commented:
Yes, the branch is now fully rebased onto the latest main and the pull request is now completely updated with all current changes.

abdeljalilsennaoui committed:
Merge branch 'feature/interactive-feedback-reviews-qa' of https://github.com/Lekerj/SOEN341 into feature/interactive-feedback-reviews-qa

abdeljalilsennaoui committed:
Database: Add average_rating column and update seed file

- Add migration script to add average_rating column to users table
- The average_rating column is critical for the Reviews & Q&A feature as the backend's recalcAverageRating() function depends on it
- Update seed_reviews_qa.sql to use COALESCE for safer NULL handling
- Add SQL migration file for documentation purposes

This fixes a schema gap where the backend code expected the average_rating column but it was missing from the users table.

abdeljalilsennaoui reviewed (approved):
Excellent implementation! All backend tests passing, components are modular and well-structured. Minor note: fixed missing buildDistributionPlaceholders() function in reviews-tab.js for the demo rendering. Ready to integrate into event detail pages.

I noticed a missing column in users tables so I made an additional script to remedy the issue: 
this PR adds the missing average_rating column to the users table, which is required by the Reviews & Q&A feature. Why: The backend's recalcAverageRating() function updates this column after every review operation (POST, PUT, DELETE) to store the organizer's average rating for performance optimization. What's included:
migrate-add-average-rating.js - Migration script to safely add the column
Updated seed file with better NULL handling
All 125 backend tests pass ✓
Setup:
mysql -h localhost -u 341 -p < backend/sql/setup-database.sql
cd backend && node migrate-add-average-rating.js && cd ..
mysql -h localhost -u 341 -p < backend/sql/seed_reviews_qa.sql  # optional
Then run:
cd backend && npm run start:local    # port 3000
cd frontend && npm run start         # port 8080

mickeyblue89 reviewed (approved):
Implementation working perfectly and responsive as expected, also aligning with other pages designs.

Also thank you Abdeljalil for adding the add column to the script as this will ensure not to miss adding the missing column manually when rebasing.

Great work overall
